passportAuth.go

```diff
@@ -0,0 +1,188 @@
+	}
+	return
+}
+
+// Close cleans up all resources
+func (p *PassportAuth) Close() error {
+	return nil
+}
+
+// Clone creates a Copy of itself
+func (p *PassportAuth) Clone() Authenticator {
+	// once created the struct is accessed read-only
+	// so there is no need to copy it
+	return &PassportAuth{
+		user:       p.user,
+		pw:         p.pw,
+		partnerUrl: p.partnerUrl,
+		cookies:    p.cookies,
+	}
+}
+
+// String toString
+func (p *PassportAuth) String() string {
+	return fmt.Sprintf("PassportAuth login: %s", p.user)
+}
+
+func (p *PassportAuth) genCookies(header *http.Header) error {
+	// For more details refer to:
+	// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-pass/2c80637d-438c-4d4b-adc5-903170a779f3
+	// Skipping step 1 and 2 as we already have the partner server challenge
+
+	baseAuthenticationServer := header.Get("Location")
+	baseAuthenticationServerURL, err := url.Parse(baseAuthenticationServer)
+	if err != nil {
+		return err
+	}
+
+	// Skipping step 3 and 4 as we already know that we need and have the user's credentials
+	// Step 5 (Sign-in request)
+	authenticationServerUrl := url.URL{
+		Scheme: baseAuthenticationServerURL.Scheme,
+		Host:   baseAuthenticationServerURL.Host,
+		Path:   "/login2.srf",
+	}
+
+	partnerServerChallenge := strings.Split(header.Get("WWW-Authenticate"), " ")[1]
+
+	req := http.Request{
+		Method: "GET",
+		URL:    &authenticationServerUrl,
+		Header: http.Header{
+			"Authorization": []string{"Passport1.4 sign-in=" + url.QueryEscape(p.user) + ",pwd=" + url.QueryEscape(p.pw) + ",OrgVerb=GET,OrgUrl=" + p.partnerUrl + "," + partnerServerChallenge},
+		},
+	}
+
+	client := &http.Client{}
+
+	rs, err := client.Do(&req)
+	if err != nil {
+		return err
+	}
+	if rs.StatusCode != 200 {
+		return fmt.Errorf("%d: Incorrect credentials", rs.StatusCode)
+	}
+
+	// Step 6 (Token Response from Authentication Server)
+	tokenResponseHeader := rs.Header.Get("Authentication-Info")
+	if tokenResponseHeader == "" {
+		return fmt.Errorf("no Authentication-Info header found")
+	}
+	tokenResponseHeaderList := strings.Split(tokenResponseHeader, ",")
+	token := ""
+	for _, tokenResponseHeader := range tokenResponseHeaderList {
+		if strings.HasPrefix(tokenResponseHeader, "from-PP='") {
+			token = tokenResponseHeader
+			break
+		}
+	}
+	if token == "" {
+		return fmt.Errorf("no token found in Authentication-Info header")
+	}
+
+	// Step 7 (First Authentication Request to Partner Server)
+	origUrl, err := url.Parse(p.partnerUrl)
+	if err != nil {
+		return err
+	}
+	req = http.Request{
+		Method: "GET",
+		URL:    origUrl,
+		Header: http.Header{
+			"Authorization": []string{"Passport1.4 " + token},
+		},
+	}
+
+	rs, err = client.Do(&req)
+	if err != nil {
+		return err
+	}
+	if rs.StatusCode != 200 && rs.StatusCode != 302 {
+		return fmt.Errorf("%d: authentication failed with Partner Server", rs.StatusCode)
+	}
+
+	// Step 8 (Set Token Message from Partner Server)
+	p.cookies = []http.Cookie{}
```

Comment:
actually we could use `len(rs.Header.Values("Set-Cookie")` as fixed initializer to and set the index instead of àppend`